File: service/service.go
package service

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"wattwarden/hal"
)

// Config represents persistent settings for WattWarden
type Config struct {
	AutoExtremeEnabled bool `json:"auto_extreme_enabled"`
	AutoBrightness     bool `json:"auto_brightness"`
}

// GetConfigPath returns the cross-platform path for the configuration file
func GetConfigPath() string {
	if runtime.GOOS == "windows" {
		programData := os.Getenv("ProgramData")
		if programData == "" {
			programData = "C:\\ProgramData"
		}
		return filepath.Join(programData, "wattwarden", "config.json")
	}
	return "/etc/wattwarden/config.json"
}

// GetPIDPath returns the path to the daemon PID file
func GetPIDPath() string {
	if runtime.GOOS == "windows" {
		programData := os.Getenv("ProgramData")
		if programData == "" {
			programData = "C:\\ProgramData"
		}
		return filepath.Join(programData, "wattwarden", "wattwarden.pid")
	}
	return "/var/run/wattwarden.pid"
}

// LoadConfig reads the configuration from disk, with default fallbacks
func LoadConfig() Config {
	cfg := Config{
		AutoExtremeEnabled: true,
		AutoBrightness:     true,
	}
	data, err := os.ReadFile(GetConfigPath())
	if err == nil {
		_ = json.Unmarshal(data, &cfg)
	}
	return cfg
}

// SaveConfig writes the configuration to disk atomically
func SaveConfig(cfg Config) error {
	p := GetConfigPath()
	if err := os.MkdirAll(filepath.Dir(p), 0755); err != nil {
		return err
	}
	data, err := json.MarshalIndent(cfg, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(p, data, 0644)
}

// IsProcessAlive checks if a process with the given PID is currently running
func IsProcessAlive(pid int) bool {
	if pid <= 0 {
		return false
	}
	process, err := os.FindProcess(pid)
	if err != nil {
		return false
	}
	if runtime.GOOS == "windows" {
		return true // FindProcess on Windows verifies process existence
	}
	// On Unix, sending signal 0 checks for process existence without killing it
	return process.Signal(syscall.Signal(0)) == nil
}

// IsDaemonActive checks if the WattWarden daemon is running in the background
func IsDaemonActive() bool {
	pidPath := GetPIDPath()
	data, err := os.ReadFile(pidPath)
	if err == nil {
		pidStr := strings.TrimSpace(string(data))
		if pid, err := strconv.Atoi(pidStr); err == nil {
			if IsProcessAlive(pid) {
				return true
			}
		}
	}

	if runtime.GOOS == "linux" {
		out, err := exec.Command("systemctl", "is-active", "wattwarden.service").Output()
		if err == nil && strings.TrimSpace(string(out)) == "active" {
			return true
		}
	}

	return false
}

// RunDaemon runs the daemon blocking loop for systemd, launchd, or background runner
func RunDaemon(b hal.Backend) {
	// Write current PID to PID file
	pidPath := GetPIDPath()
	_ = os.MkdirAll(filepath.Dir(pidPath), 0755)
	_ = os.WriteFile(pidPath, []byte(strconv.Itoa(os.Getpid())), 0644)
	defer os.Remove(pidPath)

	cfg := LoadConfig()
	b.SetAutoBrightness(cfg.AutoBrightness)
	b.StartAutoExtremeDaemon()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM, syscall.SIGHUP)
	<-sigChan

	b.StopDaemon()
}

// StartBackgroundDaemon ensures the daemon is started in background
func StartBackgroundDaemon(b hal.Backend) error {
	cfg := LoadConfig()
	cfg.AutoExtremeEnabled = true
	_ = SaveConfig(cfg)

	// If systemd is available and service is installed, use systemctl
	if runtime.GOOS == "linux" {
		if _, err := os.Stat("/etc/systemd/system/wattwarden.service"); err == nil {
			_ = exec.Command("systemctl", "restart", "wattwarden.service").Run()
			if IsDaemonActive() {
				return nil
			}
		}
	} else if runtime.GOOS == "darwin" {
		if _, err := os.Stat("/Library/LaunchDaemons/com.wattwarden.daemon.plist"); err == nil {
			_ = exec.Command("launchctl", "start", "com.wattwarden.daemon").Run()
			if IsDaemonActive() {
				return nil
			}
		}
	}

	// Also start in-process daemon
	if !b.IsDaemonRunning() {
		b.StartAutoExtremeDaemon()
	}

	// Otherwise spawn detached background daemon process
	if !IsDaemonActive() {
		_ = SpawnDetachedDaemon()
	}
	return nil
}

// StopBackgroundDaemon stops the daemon locally and any background service/process
func StopBackgroundDaemon(b hal.Backend) {
	cfg := LoadConfig()
	cfg.AutoExtremeEnabled = false
	_ = SaveConfig(cfg)

	b.StopDaemon()

	if runtime.GOOS == "linux" {
		_ = exec.Command("systemctl", "stop", "wattwarden.service").Run()
	} else if runtime.GOOS == "darwin" {
		_ = exec.Command("launchctl", "stop", "com.wattwarden.daemon").Run()
	}

	pidPath := GetPIDPath()
	data, err := os.ReadFile(pidPath)
	if err == nil {
		pidStr := strings.TrimSpace(string(data))
		if pid, err := strconv.Atoi(pidStr); err == nil {
			if proc, err := os.FindProcess(pid); err == nil {
				_ = proc.Signal(syscall.SIGTERM)
			}
		}
		_ = os.Remove(pidPath)
	}
}

// InstallService installs the auto-start background service for the current OS
func InstallService() error {
	exePath, err := os.Executable()
	if err != nil {
		exePath = "/usr/local/bin/wattwarden"
	}

	cfg := LoadConfig()
	_ = SaveConfig(cfg)

	switch runtime.GOOS {
	case "linux":
		serviceContent := fmt.Sprintf(`[Unit]
Description=WattWarden Auto Power and Hardware Management Daemon
After=multi-user.target

[Service]
Type=simple
ExecStart=%s --daemon
Restart=always
RestartSec=3
KillMode=process

[Install]
WantedBy=multi-user.target
`, exePath)

		if err := os.WriteFile("/etc/systemd/system/wattwarden.service", []byte(serviceContent), 0644); err != nil {
			return err
		}
		_ = exec.Command("systemctl", "daemon-reload").Run()
		_ = exec.Command("systemctl", "enable", "--now", "wattwarden.service").Run()
		return nil

	case "darwin":
		plistContent := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Label</key>
    <string>com.wattwarden.daemon</string>
    <key>ProgramArguments</key>
    <array>
        <string>%s</string>
        <string>--daemon</string>
    </array>
    <key>RunAtLoad</key>
    <true/>
    <key>KeepAlive</key>
    <true/>
    <key>StandardErrorPath</key>
    <string>/var/log/wattwarden.log</string>
    <key>StandardOutPath</key>
    <string>/var/log/wattwarden.log</string>
</dict>
</plist>
`, exePath)

		plistPath := "/Library/LaunchDaemons/com.wattwarden.daemon.plist"
		if err := os.WriteFile(plistPath, []byte(plistContent), 0644); err != nil {
			return err
		}
		_ = exec.Command("launchctl", "load", "-w", plistPath).Run()
		return nil

	case "windows":
		_ = exec.Command("schtasks", "/Create", "/TN", "WattWardenDaemon", "/TR", fmt.Sprintf("\"%s\" --daemon", exePath), "/SC", "ONSTART", "/RU", "SYSTEM", "/RL", "HIGHEST", "/F").Run()
		_ = exec.Command("schtasks", "/Run", "/TN", "WattWardenDaemon").Run()
		return nil
	}

	return fmt.Errorf("unsupported operating system: %s", runtime.GOOS)
}

// UninstallService removes the background service for the current OS
func UninstallService() error {
	switch runtime.GOOS {
	case "linux":
		_ = exec.Command("systemctl", "disable", "--now", "wattwarden.service").Run()
		_ = os.Remove("/etc/systemd/system/wattwarden.service")
		_ = exec.Command("systemctl", "daemon-reload").Run()
		return nil
	case "darwin":
		plistPath := "/Library/LaunchDaemons/com.wattwarden.daemon.plist"
		_ = exec.Command("launchctl", "unload", "-w", plistPath).Run()
		_ = os.Remove(plistPath)
		return nil
	case "windows":
		_ = exec.Command("schtasks", "/Delete", "/TN", "WattWardenDaemon", "/F").Run()
		return nil
	}
	return nil
}

File: crates/wattwarden-cli/src/service.rs
//! Service installation + installed-binary sync, transcribed from `service/service.go`.
//!
//! The exact systemd unit text is part of the interface contract (`PARITY.md` §3) and
//! deliberately differs from `wattwarden-daemon`'s internal helper, so it lives here in
//! the CLI crate rather than reusing that one.

use std::path::Path;
use std::process::Command;

/// Go `InstallService`: the unit always points at the installed location.
pub const INSTALLED_BINARY: &str = "/usr/local/bin/wattwarden";

#[cfg(target_os = "linux")]
const SYSTEMD_UNIT_PATH: &str = "/etc/systemd/system/wattwarden.service";

#[cfg(target_os = "macos")]
const PLIST_PATH: &str = "/Library/LaunchDaemons/com.wattwarden.daemon.plist";

/// Go `SyncInstalledBinary`: when running as root from a different path, copy the
/// current executable to `/usr/local/bin/wattwarden` with mode 0755.
pub fn sync_installed_binary() {
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;

        if nix::unistd::Uid::effective().as_raw() != 0 {
            return;
        }
        let Ok(current_exe) = std::env::current_exe() else {
            return;
        };
        if current_exe == Path::new(INSTALLED_BINARY) {
            return;
        }
        let Ok(bytes) = std::fs::read(&current_exe) else {
            return;
        };
        let _ = std::fs::create_dir_all("/usr/local/bin");
        if std::fs::write(INSTALLED_BINARY, bytes).is_ok() {
            let _ =
                std::fs::set_permissions(INSTALLED_BINARY, std::fs::Permissions::from_mode(0o755));
        }
    }
}

/// The exact systemd unit body from `PARITY.md` §3 / Go `InstallService`.
#[cfg(target_os = "linux")]
pub fn systemd_unit_text() -> String {
    format!(
        "[Unit]\n\
         Description=WattWarden Auto Power and Hardware Management Daemon\n\
         After=multi-user.target\n\
         \n\
         [Service]\n\
         Type=simple\n\
         ExecStart={INSTALLED_BINARY} --daemon\n\
         Restart=always\n\
         RestartSec=3\n\
         KillMode=process\n\
         \n\
         [Install]\n\
         WantedBy=multi-user.target\n"
    )
}

/// Go `InstallService`.
pub fn install_service() -> Result<(), String> {
    sync_installed_binary();

    #[cfg(target_os = "linux")]
    {
        let content = systemd_unit_text();

        std::fs::write(SYSTEMD_UNIT_PATH, content).map_err(|e| e.to_string())?;
        let _ = Command::new("systemctl").arg("daemon-reload").status();
        let _ = Command::new("systemctl")
            .args(["enable", "--now", "wattwarden.service"])
            .status();
        return Ok(());
    }

    #[cfg(target_os = "macos")]
    {
        let content = format!(
            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n\
             <!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n\
             <plist version=\"1.0\">\n\
             <dict>\n\
             \x20   <key>Label</key>\n\
             \x20   <string>com.wattwarden.daemon</string>\n\
             \x20   <key>ProgramArguments</key>\n\
             \x20   <array>\n\
             \x20       <string>{INSTALLED_BINARY}</string>\n\
             \x20       <string>--daemon</string>\n\
             \x20   </array>\n\
             \x20   <key>RunAtLoad</key>\n\
             \x20   <true/>\n\
             \x20   <key>KeepAlive</key>\n\
             \x20   <true/>\n\
             \x20   <key>StandardErrorPath</key>\n\
             \x20   <string>/var/log/wattwarden.log</string>\n\
             \x20   <key>StandardOutPath</key>\n\
             \x20   <string>/var/log/wattwarden.log</string>\n\
             </dict>\n\
             </plist>\n"
        );
        std::fs::write(PLIST_PATH, content).map_err(|e| e.to_string())?;
        let _ = Command::new("launchctl")
            .args(["load", "-w", PLIST_PATH])
            .status();
        return Ok(());
    }

    #[cfg(target_os = "windows")]
    {
        let _ = Command::new("schtasks")
            .args([
                "/Create",
                "/TN",
                "WattWardenDaemon",
                "/TR",
                &format!("\"{INSTALLED_BINARY}\" --daemon"),
                "/SC",
                "ONSTART",
                "/RU",
                "SYSTEM",
                "/RL",
                "HIGHEST",
                "/F",
            ])
            .status();
        let _ = Command::new("schtasks")
            .args(["/Run", "/TN", "WattWardenDaemon"])
            .status();
        return Ok(());
    }

    #[allow(unreachable_code)]
    Ok(())
}

/// Go `UninstallService`.
pub fn uninstall_service() -> Result<(), String> {
    #[cfg(target_os = "linux")]
    {
        let _ = Command::new("systemctl")
            .args(["disable", "--now", "wattwarden.service"])
            .status();
        let _ = std::fs::remove_file(SYSTEMD_UNIT_PATH);
        let _ = Command::new("systemctl").arg("daemon-reload").status();
        return Ok(());
    }

    #[cfg(target_os = "macos")]
    {
        let _ = Command::new("launchctl")
            .args(["unload", "-w", PLIST_PATH])
            .status();
        let _ = std::fs::remove_file(PLIST_PATH);
        return Ok(());
    }

    #[cfg(target_os = "windows")]
    {
        let _ = Command::new("schtasks")
            .args(["/Delete", "/TN", "WattWardenDaemon", "/F"])
            .status();
        return Ok(());
    }

    #[allow(unreachable_code)]
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The unit text is part of the contract (`PARITY.md` §3): lock it down exactly,
    /// including `Restart=always`, `RestartSec=3`, `KillMode=process` and the
    /// `--daemon` argument.
    #[cfg(target_os = "linux")]
    #[test]
    fn systemd_unit_text_matches_parity_contract() {
        let expected = "[Unit]\n\
Description=WattWarden Auto Power and Hardware Management Daemon\n\
After=multi-user.target\n\
\n\
[Service]\n\
Type=simple\n\
ExecStart=/usr/local/bin/wattwarden --daemon\n\
Restart=always\n\
RestartSec=3\n\
KillMode=process\n\
\n\
[Install]\n\
WantedBy=multi-user.target\n";

        assert_eq!(systemd_unit_text(), expected);
    }
}
